File: leetcode/easy/144-preorder/stack.cpp
/**
 * @file sol.cpp
 * @author Vortexx2
 * @brief Problem 144 - Binary Tree Preorder Traversal
 * @date 20-07-2021
 * 
 * Runtime - 4 ms
 * Memory Usage - 8.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  /**
   * @brief Using a stack to store addresses of previous nodes where tree branched.
   * 
   * @param root 
   * @return vector<int> 
   */
  vector<int> preorderTraversal(TreeNode *root)
  {
    stack<TreeNode *> st;
    vector<int> res;

    while (true)
    {

      while (root)
      {
        res.push_back(root->val);

        st.push(root);
        root = root->left;
      }

      if (st.empty())
        break;

      root = st.top();
      st.pop();
      root = root->right;
    }

    return res;
  }
};

File: leetcode/easy/144-preorder/stack.py
"""
  @author Vortexx2

  Runtime - 24 ms
  Memory Usage - 14.2 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def preorderTraversal(self, root: TreeNode) -> List[int]:
    st = []
    ans = []

    while True:

      while root:
        ans.append(root.val)
        st.append(root)
        root = root.left

      if not len(st):
        break

      root = st.pop().right

    return ans

